ocaml/xapi/repository.ml

```diff
@@ -0,0 +1,975 @@
+            ignore (Helpers.call_script cmd_modifyrepo
+                      ["--remove"; "updateinfo"; repo_dir ^ "/repodata"]);
+            ignore (Helpers.call_script cmd_modifyrepo
+                      ["--mdtype"; "updateinfo"; xml_file_path; repo_dir ^ "/repodata"]))
+      | false -> raise Api_errors.(Server_error (no_updateinfo_xml, []))
+    end
+  | false ->
+    error "local pool repository directory %s does not exists" !Xapi_globs.local_pool_repo_dir;
+    raise Api_errors.(Server_error (no_local_pool_repository, []))
+
+let with_pool_repository ~__context ~self ~prefix f =
+  Xapi_stdext_pervasives.Pervasiveext.finally
+    (fun () ->
+        Mutex.lock exposing_pool_repo_mutex;
+        f ())
+    (fun () -> Mutex.unlock exposing_pool_repo_mutex)
+
+let is_local_pool_repo_enabled () =
+  match Mutex.try_lock exposing_pool_repo_mutex with
+  | true ->
+    Mutex.unlock exposing_pool_repo_mutex;
+    false
+  | false -> true
+
+let reposync_try_lock () =
+  Mutex.try_lock reposync_mutex
+
+let reposync_unlock () =
+  Mutex.unlock reposync_mutex
+
+let with_reposync_lock f =
+  if Mutex.try_lock reposync_mutex then
+    Xapi_stdext_pervasives.Pervasiveext.finally
+      (fun () -> f ())
+      (fun () -> Mutex.unlock reposync_mutex)
+  else
+    raise Api_errors.(Server_error (reposync_in_progress, []))
+
+let get_local_repomd_xml_path ~__context ~self ~prefix =
+  let name = Db.Repository.get_name_label ~__context ~self in
+  Printf.sprintf "%s/%s-%s/repodata/repomd.xml" !Xapi_globs.local_pool_repo_dir prefix name
+
+let try_set_available_updates ~__context ~self ~prefix =
+  let hosts = Db.Host.get_all ~__context in
+  let xml_path = get_local_repomd_xml_path ~__context ~self ~prefix in
+  let md = parse_repomd xml_path in
+  let hash = md.checksum in
+  let not_up_to_date host () =
+    let json = http_get_host_updates_in_json ~__context ~host ~installed:false in
+    (* TODO: live patches *)
+    match Yojson.Basic.Util.member "updates" json with
+    | `List [] -> false
+    | _ -> true
+    | exception e ->
+      let uuid = Db.Host.get_uuid ~__context ~self:host in
+      error "Invalid updates from host uuid='%s': %s" uuid (ExnHelper.string_of_exn e);
+      raise Api_errors.(Server_error (get_host_updates_failed, [uuid]))
+  in
+  let funs = List.map (fun h -> not_up_to_date h) hosts in
+  let rets_of_hosts = run_in_parallel funs capacity_in_parallel in
+  let is_all_up_to_date = not (List.exists (fun b -> b) rets_of_hosts) in
+  Db.Repository.set_up_to_date ~__context ~self  ~value:is_all_up_to_date;
+  Db.Repository.set_hash ~__context ~self ~value:hash;
+  hash
+
+let update_of_json j =
+  let open Yojson.Basic.Util in
+  let u = {
+    name = (member "name" j |> to_string);
+    arch = (member "arch" j |> to_string);
+    new_ver_rel = (member "newVerRel" j |> to_string);
+    old_ver_rel = (try Some (member "oldVerRel" j |> to_string) with _ -> None);
```

Comment:
I don't think the versions should just be string.

I think this is where they should be parsed into `{version, release}`

For example:
```
let ver_rel_of_string str = match Astring.cut ~sep:"-" with
| Some (v, r) when v <> "" && r <> "" ->
  {version=v; release=r}
| None ->
  None


let new_ver_rel = (member "newVerRel" j |> to_string |> ver_rel_of_string)